src/main/java/com/github/jikoo/regionerator/ChunkFlagger.java

```diff
@@ -6,120 +6,79 @@
-	private BukkitTask saveTask = null;
+	private final LoadingCache<String, FlagData> flagFileCache;
+	private final Connection database;
 
 	ChunkFlagger(Regionerator plugin) {
 		this.plugin = plugin;
-		this.saveQueue = new HashMap<>();
+		try {
+			Class.forName("org.sqlite.JDBC").newInstance();
+			this.database = DriverManager.getConnection("jdbc:sqlite://" + plugin.getDataFolder().getAbsolutePath() + "/data.db");
+			try (Statement st = database.createStatement()) {
+				st.executeUpdate("CREATE TABLE IF NOT EXISTS \"chunkdata\"(\"chunk_id\" TEXT NOT NULL UNIQUE, \"time\" INTEGER NOT NULL)");
+			}
+		} catch (Exception e) {
+			e.printStackTrace();
+			throw new RuntimeException("An error occurred while connecting SQLite");
+		}
 		this.flagFileCache = CacheBuilder.newBuilder()
 				// Minimum 1 minute cache duration even if flags are saved more often than every 30s
 				.expireAfterAccess(Math.max(60, plugin.getTicksPerFlagAutosave() * 10), TimeUnit.SECONDS)
-				.removalListener(new RemovalListener<Pair<String, String>, Pair<YamlConfiguration, Boolean>>() {
-					@Override
-					public void onRemoval(@NotNull RemovalNotification<Pair<String, String>, Pair<YamlConfiguration, Boolean>> notification) {
-						// Only attempt to save if dirty to minimize writes
-						if (notification.getValue().getRight()) {
-							saveQueue.put(notification.getKey(), notification.getValue().getLeft());
-							startSaveQueue();
+				.removalListener((RemovalListener<String, FlagData>) notification -> {
+					// Only attempt to save if dirty to minimize writes
+					if (notification.getValue().dirty) {
+						try (Statement st = database.createStatement()) {
+							st.executeUpdate("INSERT OR REPLACE INTO \"chunkdata\"(\"chunk_id\", \"time\") VALUES ('" + notification.getValue().chunkId + "', '" + notification.getValue().time + "')");
+						} catch (Exception e) {
+							e.printStackTrace();
 						}
 					}
-				}).build(new CacheLoader<Pair<String, String>, Pair<YamlConfiguration, Boolean>>() {
+				}).build(new CacheLoader<String, FlagData>() {
 					@Override
-					public Pair<YamlConfiguration, Boolean> load(@NotNull Pair<String, String> key) throws Exception {
-						if (saveQueue.containsKey(key)) {
-							return new Pair<>(saveQueue.get(key), false);
+					public FlagData load(@NotNull String key) throws Exception {
+						try (Statement st = database.createStatement(); ResultSet rs = st.executeQuery("SELECT \"time\" FROM \"chunkdata\" WHERE \"chunk_id\" = '" + key + "'")) {
+							if (rs.next()) {
+								return new FlagData(key, rs.getLong(1));
+							} else {
+								return new FlagData(key, -1);
+							}
 						}
-						File flagFile = getFlagFile(key);
-						if (flagFile.exists()) {
-							return new Pair<>(YamlConfiguration.loadConfiguration(flagFile), false);
-						}
-						return new Pair<>(new YamlConfiguration(), false);
 					}
 				});
 
 		convertOldFlagsFile();
-	}
-
-	private void startSaveQueue() {
```

Comment:
I see you've removed the queue system, but the reason it was added is that the Cache has a tendency to invalidate large batches simultaneously. Is it possible to perform batch updates with SQLite? I know it's pretty easy with SQL.

Reply:
I’m not so good at SQL, but I found some information and experimented, I found out it is possible to significantly increase write performance if you turn off instant data saving by: "database.setAutoCommit(false)" and create a task that will save from time to time data by "database.commit()", batch in this case, doesn’t reduce the time spent on multiple insertion too much, about 5%

Reply:
Essentially that, there are a couple of other calls that should simplify what data gets sent to the SQL server.
`PreparedStatement#addBatch` should flag a statement as part of a batch update, `PreparedStatement#executeBatch` on the last statement when ready, then `Connection#commit`.